specification/transport/udp/udp.tex

```diff
@@ -0,0 +1,370 @@
+# Incremented by one for each subsequent frame of a multi-frame transfer.
+
+bool end_of_transfer
+# True if this is the last frame of a multi-frame transfer, or a single-frame transfer.
+
+uint16 user_data
+# Opaque application-specific data with user-defined semantics.
+# Generic implementations should emit zero and ignore this field upon reception.
+
+uint8[2] header_crc16_big_endian
+# CRC-16/CCITT-FALSE of the preceding serialized header data in the big endian byte order.
+# Application of the CRC function to the entire header shall yield zero, otherwise the header is malformed.
+
+@assert _offset_ / 8 == {24}
+@sealed     # The payload data follows.
+\end{minted}
+\end{samepage}
+
+The header CRC function is \textbf{CRC-16/CCITT-FALSE};
+refer to section~\ref{sec:appendix_crc16ccitt_false} for further information.
+
+\begin{remark}
+    Certain states provided in the header duplicate information that is already available in the IP header
+    or the multicast group address.
+    This is done for reasons of unification of the header format with other standard transport layer definitions,
+    and to simplify the access to the transfer parameters that otherwise would be hard to reach above the
+    network layer, such as the DSCP value.
+    The latter consideration is particularly important for forwarding nodes.
+\end{remark}
+
+\subsection{Transfer payload}
+
+After the transfer payload is constructed but before it is scheduled for transmission over the network,
+it is appended with the transfer CRC field.
+The transfer CRC function is \textbf{CRC-32C} (section~\ref{sec:appendix_crc32c}),
+and its value is serialized in the little-endian byte order.
+The transfer CRC function is applied to the entire transfer payload and only transfer payload.
+
+The transfer CRC is provided for all transfers,
+including single-frame transfers and transfers with an empty payload\footnote{%
+    This provides end-to-end integrity protection for the transfer payload.
+}.
+An implementation receiving a transfer should verify the correctness of its transfer CRC.
+
+\begin{remark}
+    From the perspective of the multi-frame segmentation logic, the transfer CRC field is part of the transfer payload.
+    From the definition of the header format it follows that the transfer CRC can only be found at the end of
+    the packet if the \verb|end_of_transfer| bit is set,
+    unless the transfer CRC field has spilled over to the next frame
+    (in which case the frame would contain only the transfer CRC itself or the tail thereof).
+\end{remark}
+
+\subsection{Maximum transmission unit}
+
+In this section, the maximum transmission unit (MTU) is defined as the maximum size of a UDP/IP datagram payload.
+
+This specification does not restrict the MTU of the underlying transport.
+It is recommended, however, to avoid MTU values less than 508~bytes,
+allowing applications to exchange up to $508 - 24 - 4 = 480$ bytes of payload in a single-frame transfer.
```

Comment:
Should we recommend 1280 octets since that's the IPV6 link minimum (per https://datatracker.ietf.org/doc/html/rfc8200)? I'm not sure what the general MTU support available is for networking equipment but looking at my Macbook, for example, it lists 1500 as "standard" MTU for both a USB ethernet link I have and the internal wifi radio.

Reply:
Indeed, 1500 bytes is standard for Ethernet, and I am fairly confident that 1280 will work anywhere unless you want to shuttle your packet across the public Internet, but there is another concern we should be considering: the worst-case preemption latency, which is a function of the link speed and MTU. You may remember that AFDX limits the MTU per VL type for this reason. By requiring (or recommending) a higher MTU we may strip the designer of the flexibility required to meet the timing constraints. You might argue here that this is less relevant for Cyphal/UDP as there is Cyphal/TSN on the horizon, but unless there are strong reasons to do otherwise, we should not undermine the ability of the protocol to meet stricter real-time requirements.